buddy_init/init_framework.py

```diff
@@ -0,0 +1,97 @@
+import argparse
+import inspect
+import os
+from typing import Callable, Union, Dict, Any
+from shutil import which
+from functools import wraps
+from halo import Halo
+
+
+class FatalException(Exception):
+    pass
+
+
+def log_output(fn):
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        with open('/tmp/buddy-init.log', 'a'):
+            return fn(*args, **kwargs)
+
+    return wrapper()
+
+
+def preflight_check():
+    if which('ssh') is None:
+        raise FatalException('missing ssh, you can install it with "sudo apt install openssl-client"')
+    if which('git') is None:
+        raise FatalException('missing git, you can install it with "sudo apt install git"')
+    if which('pip') is None:
+        raise FatalException('missing pip, you can install it with "sudo apt install python-pip"')
+
+
+def coalesce(*args):
+    return next((i for i in args if i is not None), None)
+
+
+def get_fn_parameters(action):
+    if hasattr(action, 'func'):
+        return list(set(inspect.signature(action).parameters) - set(action.keywords.keys()))
+    else:
+        return list(inspect.signature(action).parameters)
+
+
+class WorkingDirectory:
+
+    def __init__(self, *, path, force):
+        if force:
+            os.system(f"rm -rf {path}")
+        elif os.path.exists(path) and os.listdir(path):
+            raise Exception("The directory exists and it's not empty")
+
+        os.makedirs(path)
+        self.base_path = path
+        self.current_path = os.getcwd()
+
+    def __enter__(self):
+        os.chdir(self.base_path)
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        os.chdir(self.current_path)
+
+
+def prompt(action: Union[Callable[[], None], Callable[[str], None]],
```

Comment:
Type or not type but be consistent

Reply:
Typed everything I could type

Reply:
:cookie:

Reply:
ahah there is still type to be typed. No pressure, i won't mention it more until we have a linter.